azure-iot-hub/tests/test_iothub_registry_manager.py

```diff
@@ -0,0 +1,800 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""This module should be replaced when there are actual tests for the azure.iot.hub library"""
+
+import pytest
+from azure.iot.hub.iothub_registry_manager import IoTHubRegistryManager
+from azure.iot.hub.protocol.iot_hub_gateway_service_ap_is20190630 import (
+    IotHubGatewayServiceAPIs20190630,
+)
+
+from azure.iot.hub.auth import ConnectionStringAuthentication
+
+from azure.iot.hub.protocol.models import AuthenticationMechanism
+
+
+fake_shared_access_key = "Zm9vYmFy"
+fake_shared_access_key_name = "alohomora"
+
+fake_primary_key = "petrificus"
+fake_secondary_key = "totalus"
+fake_primary_thumbprint = "HELFKCPOXAIR9PVNOA3"
+fake_secondary_thumbprint = "RGSHARLU4VYYFENINUF"
+fake_hostname = "beauxbatons.academy-net"
+fake_device_id = "MyPensieve"
+fake_module_id = "Divination"
+fake_managed_by = "Hogwarts"
+fake_etag = "taggedbymisnitryogmagic"
+fake_status = "flying"
+
+
+@pytest.fixture(scope="function")
+def mock_service_operations(mocker):
+    mock_service_operations_init = mocker.patch(
+        "azure.iot.hub.protocol.iot_hub_gateway_service_ap_is20190630.ServiceOperations"
+    )
+    return mock_service_operations_init.return_value
+
+
+@pytest.fixture(scope="function")
+def mock_device_constructor(mocker):
+    return mocker.patch("azure.iot.hub.iothub_registry_manager.Device")
+
+
+@pytest.fixture(scope="function")
+def mock_module_constructor(mocker):
+    return mocker.patch("azure.iot.hub.iothub_registry_manager.Module")
+
+
+@pytest.fixture(scope="function")
+def iothub_registry_manager():
+    connection_string = "HostName={hostname};DeviceId={device_id};SharedAccessKeyName={skn};SharedAccessKey={sk}".format(
+        hostname=fake_hostname,
+        device_id=fake_device_id,
+        skn=fake_shared_access_key_name,
+        sk=fake_shared_access_key,
+    )
+    iothub_registry_manager = IoTHubRegistryManager(connection_string)
+    return iothub_registry_manager
+
+
+@pytest.mark.describe("IoTHubRegistryManager")
+class TestIoTHubRegistryManagerInit(object):
+    @pytest.mark.it("instantiates correctly with auth")
+    def test_init_auth(self, iothub_registry_manager):
+        assert iothub_registry_manager is not None
+        assert iothub_registry_manager.auth is not None
+        assert isinstance(iothub_registry_manager.auth, ConnectionStringAuthentication)
+
+    @pytest.mark.it("instantiates correctly with protocol client with correct Service API version")
+    def test_init_protocol_client(self, iothub_registry_manager):
+        assert iothub_registry_manager is not None
+        assert iothub_registry_manager.protocol is not None
+        assert isinstance(iothub_registry_manager.protocol, IotHubGatewayServiceAPIs20190630)
+
+
+different_devices_sas = [
```

Comment:
Using structures like this to parametrize tests is an antipattern that we are trying to avoid now due to inconsistency in output and difficulty of long term maintenance.

Parametrization should only be used for inputs to a unit under test, not for structuring tests themselves.

Reply:
I agree strongly.  This is a pattern I started and I regret it.  :(

---
In reply to: [331228799](https://github.com/Azure/azure-iot-sdk-python/pull/340#discussion_r331228799) [](ancestors = 331228799)